src/lib.rs

```diff
@@ -262,20 +262,125 @@ impl<T> StableVec<T> {
     /// has a worst case time complexity of O(n). If you already know the
     /// index, use [`remove()`](#method.remove) instead.
     pub fn pop(&mut self) -> Option<T> {
-        let last_index = self.deleted
-            .iter()
-            .enumerate()
-            .rev()
-            .find(|&(_, deleted)| !deleted)
-            .map(|(i, _)| i)
-            .unwrap_or(0);
-        self.remove(last_index)
+        self.find_last_index().and_then(|index| self.remove(index))
+    }
+
+    /// Removes and returns the first element from this collection, or `None` if
+    /// it's empty.
+    ///
+    /// This method uses exactly the same deletion strategy as
+    /// [`remove()`](#method.remove).
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use stable_vec::StableVec;
+    /// let mut sv = StableVec::from(&[1, 2, 3]);
+    /// assert_eq!(sv.pop_first(), Some(1));
+    /// assert_eq!(sv.into_vec(), vec![2, 3]);
+    /// ```
+    ///
+    /// # Note
+    ///
+    /// This method needs to find index of the first valid element. Finding it
+    /// has a worst case time complexity of O(n). If you already know the
+    /// index, use [`remove()`](#method.remove) instead.
+    pub fn pop_first(&mut self) -> Option<T> {
+        self.find_first_index().and_then(|index| self.remove(index))
+    }
+
+    /// Finds the first element and returns a reference to it, or `None` if it's empty.
+    ///
+    /// This method has O(n) complexity.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use stable_vec::StableVec;
+    /// let mut sv = StableVec::from(&[1, 2]);
+    /// sv.remove(0);
+    /// assert_eq!(sv.find_first(), Some(&2));
+    /// ```
+    pub fn find_first(&self) -> Option<&T> {
+        self.find_first_index().map(|index| &self.data[index])
+    }
+
+    /// Finds the first element and returns it's index, or `None` if it's empty.
+    ///
+    /// This method has O(n) complexity.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use stable_vec::StableVec;
+    /// let mut sv = StableVec::from(&[1, 2]);
+    /// sv.remove(0);
+    /// assert_eq!(sv.find_first_index(), Some(1));
+    /// ```
+    pub fn find_first_index(&self) -> Option<usize> {
+        match self.used_count {
+            0 => None,
+            _ => self.deleted
+                .iter()
+                .enumerate()
+                .find(|&(_, deleted)| !deleted)
+                .map(|(i, _)| i),
+        }
+    }
+
+    /// Finds the last element and returns a reference to it, or `None` if it's empty.
+    ///
+    /// This method has O(n) complexity.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use stable_vec::StableVec;
+    /// let mut sv = StableVec::from(&[1, 2]);
+    /// sv.remove(1);
+    /// assert_eq!(sv.find_last(), Some(&1));
+    /// ```
+    pub fn find_last(&self) -> Option<&T> {
+        self.find_last_index().map(|index| &self.data[index])
+    }
```

Comment:
I would rather have this function after `find_first`, so that the order of functions is:

- `find_first`
- `find_last`
- `find_first_index`
- `find_last_index`

Or do you have a reason why it's in this other order?

Reply:
I just grouped them by name, but grouping them by functionality, as you suggested, makes more sense.